### DISCUSSION TITLE
Menu: Consider alternative to deprecating links in menu components

### DISCUSSION THREAD

**Discussion by tommie** [primefaces/community#556] [PrimeVue]
Moved from https://github.com/primefaces/primevue/issues/4633, as suggested by @tugcekucukoglu.

* *Updated with ideas from [@deguich](https://github.com/orgs/primefaces/discussions/556#discussioncomment-8132659)*

### Describe the feature you would like to see added

Based on deprecation messages, PrimeVue is deprecating the use of `<router-link>` magic inside the various menu components. Instead, it encourages the users to use template customization to do the same thing. I think this is a step back in DevEx, because it makes getting started much harder. I now have to understand how the customization API works, instead of just knowing what `to` does in Vue Router.

The amount of code to do what PrimeVue does now is non-trivial:

https://github.com/primefaces/primevue/blob/52d7d11235accc6032d05100a71c38cdca1f0a24/components/lib/menu/Menuitem.vue#L15-L28

Having to do this not only for `Menuitem`, but also e.g. `PanelMenu` makes it ever more daunting:

https://github.com/primefaces/primevue/blob/52d7d11235accc6032d05100a71c38cdca1f0a24/components/lib/panelmenu/PanelMenuSub.vue#L20-L37

### Describe the solution you'd like

I would suggest the `PrimeVue.use` function takes `linkComponent` as an option: https://github.com/primefaces/primevue/blob/52d7d11235accc6032d05100a71c38cdca1f0a24/components/lib/config/PrimeVue.js#L4

This would be a dynamic component that `Menuitem` et al. instantiate using `<component is>`, a mechanism already in use there for enabling the `item` slot: https://github.com/primefaces/primevue/blob/52d7d11235accc6032d05100a71c38cdca1f0a24/components/lib/panelmenu/PanelMenuSub.vue#L38

Then, the link component API would have to be documented (likely simply just a subset of `<router-link>`) and we can elect to use `<router-link>` or `<nuxt-link>` or something custom.

The `MenuItem` et al. would still contain the "interior" nodes such as icon, label, selection/highlight styling, so that their individual components have control over this. I.e. the link component only handles interaction (and optionally styling based on deep scoped styles.) Since this reduces the API from dealing with both interactivity and presentation, it makes the API smaller (with increased stability) than if the custom link component has to handle everything, in all situations.

The default link component would simply be an `<a>` element that handles the action props. It warns if it receives the router-specific `to` attribute, perhaps hinting at configuring `<router-link>`. I.e. this would be the `<a>` half of what `Menuitem` currently defaults to.

This means there is no PrimeVue NPM dependency on `vue-router` (https://github.com/primefaces/primevue/issues/4633#issuecomment-1827780322).

### Describe alternatives you have considered

* The first proposal suggested that the entire link component (including icon, label, and styling) was moved to the custom link component, however @deguich contributed that having PrimeVue's components dictate styling and layout is beneficial.
* Defining custom components for menu items also requires thinking about expanded/collapsed and selected/highlighted theming. It would be much nicer if PrimeVue know about the concept of links, even if it didn't know the router specifics. I'm guessing the number of slot props will be ever increasing to handle this. It would be much easier with a defined link API.

### Additional context

Related to #3884, #4351.

#3118 could perhaps make use of the same defined link API, though as I just added there, I think that issue is more about the theme API.

**Comment by Jupakabra** [comment]
The alternative seems to be to use the `command` handler. But I think people forget that user experience gets ruined without the href.
Having url be a `string | func` type just like the label might solve a lot of the issues, what do you think?

**Comment by tommie** [comment]
> Having url be a string | func type just like the label might solve a lot of the issues, what do you think?

That doesn't solve the problem, that PrimeFaces wants to completely remove the URL handling, to avoid the dependency to `vue-router`. The question is *what* is handling the string URL. (Also supporting vue-router `RawLocationRecord` would be nice, not just a string URL.)

**Comment by vinodkola1337** [comment]
I agree with @tommie, the deprecation of <router-link> is a step backward for developer's experience.

**Comment by mllull** [comment]
@tommie I've an App that has about 50 Vue-Views that all of those uses `Breadcrumbs`, another component affected by this breaking change.

What I had done was to "encapsulate" Primevue's Breadcrumbs into my own component, that looks like this response: https://github.com/primefaces/primevue/issues/4739#issuecomment-1814990299.

Once done, I had to replace all of the native imports of Breadcrumbs to my new component with some magic from VSCode and, volià! 

I know that this is not a pretty trick, but it works. 

Btw, I think the same as @vinodkola1337, this is a step backward because this kind of components are heavily related with a router.

**Comment by jf908** [comment]
I think this is a pretty good solution, it would be an improvement over what we have now and much better than deprecating links entirely. Having a necessary global config option like this might be a bit confusing to PrimeVue beginners but I think this is a small price to pay if the alternative is mandate all of the template code .

I think the fact that PrimeVue's MenuItem type is used throughout so many of its components is great library design because it makes it so easy to try out different styles of menus while keeping (mostly) the same data structure. I think it would be disastrous to lose that flexibility by forcing you to mess around with templates just to get links to work since there are so many Menu based components, it would take a lot of effort to encapsulate all of them.

**Comment by tve** [comment]
As a new PrimeVue user the non-support of vue-router is pretty off-putting. I get it that having vue-router as a direct dependency is not desirable but there's gotta be some simple solution so new users like me don't have to spend a couple of hours digging through the details for what needs to go into the templates.

**Comment by mseeley** [comment]
I was surprised by this change too. It was extra surprising because `to:` was deprecated and removed yet code using `to:` doesn't cause a type error because of the `[key: string]: any;`. Ick, unhelpful.

Using `command` isn't desirable. Instead using `vue-router` to resolve route options into a `url` seems to work.

From:

```
const menuItem = {
  to: { name: 'foo', params: { a: 1, b: 2 }
}
```

To:

```
const router = useRouter()
const menuItem = {
  url: router.resolve({ name: 'foo', params: { a: 1, b: 2 }).href
}
```

**Reply by deguich** [reply]
This solution will reload all the web app, this is unfortunately not like router.push (not SPA compatible).

**Reply by mseeley** [reply]
Very true. Thank you for the kind correction. 👍

**Comment by deguich** [comment]
I think the solution involving the use of the 'item' slot is not the right one either. It requires redefining the HTML code for each item. Through this slot, we are overloading the source code of PrimeVue components, while we want to use the original display and basic features of the component! For each MenuItem, I have to go check the source code and adapt it to maintain the original functionality of the component. Example for the Breadcrumb component:
```html
<a :href="item.url || '#'" :class="cx('action')" :target="item.target" :aria-current="isCurrentUrl()" @click="onClick" v-bind="ptm('action', ptmOptions)">
  <component v-if="templates && templates.itemicon" :is="templates.itemicon" :item="item" :class="cx('icon', ptmOptions)" />
  <span v-else-if="item.icon" :class="[cx('icon'), item.icon]" v-bind="ptm('icon', ptmOptions)" />
  <span v-if="item.label" :class="cx('label')" v-bind="ptm('label', ptmOptions)">{{ label() }}</span>
</a>
```
I need to update "a" tag to router-link, redefine onClick function and I'm not sure I can deal with templates.itemIcon, ptm, cx ... 
So, since I can't easily create a generic component, I simplify it for my use, and I am certain that it will cause issues in the future during an update of PrimeVue or when I want to use one of the MenuItem features that I haven't re-implemented.

Would creating a MenuItemRouter slot be a solution? If this slot is defined, it would replace the 'a' tags and contain the content of the MenuItem:
```
<component :is="templates.menuitemrouter">
  <component v-if="templates && templates.itemicon" :is="templates.itemicon" :item="item" :class="cx('icon', ptmOptions)" />
  <span v-else-if="item.icon" :class="[cx('icon'), item.icon]" v-bind="ptm('icon', ptmOptions)" />
  <span v-if="item.label" :class="cx('label')" v-bind="ptm('label', ptmOptions)">{{ label() }}</span>
</component>
```
I don't know if this solution is feasible. The goal is to leave the sole responsibility to the developer for creating the link with the routing tool they use.

**Reply by tommie** [reply]
> Would creating a MenuItemRouter slot be a solution? 

IIUC, you want the slot to be style-agnostic, and only handle the interaction aspects. That's a great conclusion. It reduces boilerplate, makes for a smaller, more stable API, and would likely interact better with themes in practice.

If we combine my proposal (global configuration) with yours (styleless), we'd end up with a `MenuItem` like

```html
<component :is="this.$primevue.config.linkComponent">
  <component v-if="templates && templates.itemicon" :is="templates.itemicon" :item="item" :class="cx('icon', ptmOptions)" />
  ...
</component>
```

where `config.linkComponent` defaults to `<a>` and can be configured during plugin installation:

```js
import MyLink from ...;
app.use(PrimeVue, { linkComponent: MyLink });
```

I find it really practical to have a single place where the router link component is defined. Having to write

```html
<Menu>
  <template #item v-slot="slotprops">
    <MyItem ...slotprops>
      <component ...slotprops />
    </template>
  </template>
</Menu>
```

everywhere makes for fragile code in projects with more than one developer. (Even more so if `MyItem` needs to handle styling, and not just interactivity.)

**Reply by tommie** [reply]
I've updated the proposal to include this idea.

**Comment by BlakeTheAwesome** [comment]
Another option that would decouple the styling from `vue-router`, would be to have an optional callback function on the menu itself, so `MenuItems` with a `to` (or `url`) would be rendered with an `<a>`, but clicking on them would trigger your callback allowing you to do a `vue-router` push, ideally using the data in your `to` object.

**Reply by jf908** [reply]
I’m not sure this would work unless you limited it to only strings as the conversion from to object to url string is library dependent and it’s important that the <a> tag has a href string attribute.

**Reply by BlakeTheAwesome** [reply]
I think the simplest solution to this would just be to provide the `url` as well, generating it using whichever library solution you want.
It would basically mean the url gives you the anchor/href functionality, but the `to` field would give you more control in an SPA app.
